infrastructure/cloudfrontLambdaAssociations.ts

```diff
@@ -0,0 +1,59 @@
+// Copyright 2016-2021, Pulumi Corporation.  All rights reserved.
+
+import * as aws from "@pulumi/aws";
+import { LambdaEdge } from "./lambdaEdge";
+
+/**
+ * Returns the lambda function associations for the cloudfront distribution.
+ * @param addSecurityHeaders If true, will create a Lambda@Edge that sets security headers on
+ * origin responses and adds the lambda to the list of associated triggers.
+ */
+export function getLambdaFunctionAssociations(addSecurityHeaders: boolean):
+    aws.types.input.cloudfront.DistributionDefaultCacheBehaviorLambdaFunctionAssociation[] | undefined {
+    if (!addSecurityHeaders) {
+        // We only need to setup one lambda function right now, so if that's not requested
+        // simply return.
+        return;
+    }
+
+    const provider = new aws.Provider("usEast1", {
+        region: aws.Region.USEast1,
+    });
+    const lambdaEdge = new LambdaEdge("security", {
+        func: getSecurityHeadersLambdaCallback(),
+        funcDescription: "Lambda function that sets security headers on a Cloudfront origin response.",
+    // tslint:disable-next-line:align
+    }, { provider });
+
+    return [{
+        includeBody: false,
+        lambdaArn: lambdaEdge.getLambdaEdgeArn(),
+        // Origin response is a type of trigger that runs the function only when CF
+        // contacts the origin for a file. Thus reducing the number of times the func
+        // is actually reduced.
+        //
+        // See the following link for all trigger events supported by CF:
+        // https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-cloudfront-trigger-events.html
+        eventType: "origin-response",
```

Comment:
I wanted to expand on this a bit. CloudFront supports different trigger types, but the one I have used here is [`origin-response`](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-cloudfront-trigger-events.html). With this trigger type it is important to note that the Lambda@Edge function runs only if CF forwards the request to the origin (S3). For cached files, the function is not executed. This also means that we won't serve the security headers on old objects that the various browsers might be requesting. Over time those objects will fall out of cache anyway due to TTL. We can change this to `viewer-response` so that it runs regardless of a file being cached as long as the response is a 200.